### PR TITLE
Fix encrypted fields bulk_update corrupting data

### LINKED ISSUE #4359: rotate-db-key corrupts data
There are two issues:
1. Our encrypted JSON/text fields do not work with bulk_update.
2. EncryptedJSONField uses repr and eval for transformations which is not the exact inverse.

### PR DESCRIPTION
**Pull request by mdellweg**
fixes #4359

### CODE REVIEW

File: pulpcore/app/models/fields.py [hunk 1]
```diff
@@ -124,24 +129,37 @@ def encrypt(self, value):
...
 
-        return eval(force_str(_fernet().decrypt(force_bytes(value))))
+        dec_value = force_str(_fernet().decrypt(force_bytes(value)))
+        try:
+            return json.loads(dec_value, cls=self.decoder)
+        except json.JSONDecodeError:
+            deprecation_logger.info(
+                "Failed to decode json in an EncryptedJSONField. Falling back to eval. "
+                "Please run pulpcore-manager rotate-db-key to repair."
+                "This is deprecated and will be removed in pulpcore 3.40."
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/app/models/fields.py:147]: you mean while running it? Yes, maybe.

File: pulpcore/app/models/fields.py [hunk 2]
```diff
@@ -124,24 +129,37 @@ def encrypt(self, value):
...
+            return eval(dec_value)
 
-    def get_db_prep_save(self, value, connection):
-        value = self.encrypt(value)
-        return super().get_db_prep_save(value, connection)
+    def get_prep_value(self, value):
+        value = super().get_prep_value(value)
+        if value is not None:
+            if hasattr(value, "as_sql"):
+                return value
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/app/models/fields.py:155]: When using bulk_update, django creates some huge `Cast(Case(When(pk=Value(1234-1234), then=Value("password")), Else(Value(None)))` like statements. We casted them to strings encrypted them and saved to db.

File: pulpcore/app/models/fields.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdellweg) [reply pulpcore/app/models/fields.py:147]: But only once...

[the remark this answers is not in the document]
Comment (gerrod3) [reply pulpcore/app/models/fields.py:147]: Alright, I guess it isn't that big of a deal.

**Review by gerrod3 — APPROVED** [review]

**Comment by mdellweg** [comment]
I can only assume this was not an issue back in Django3 when we added the key-rotation feature and a migration in pulp-ansible using bulk_update.

**Review by ggainey — APPROVED** [review]

**Comment by mdellweg** [comment]
> Backport to 3.23: 💔 cherry-picking failed — conflicts found

https://github.com/pulp/pulpcore/pull/4366

**Pull request merged**